// @mr-hope/vuepress-plugin-components/lib/client/BackToTop.js
import Vue from "vue";
import debounce from "lodash.debounce";
import { componentLocales } from "./define";
let scrollHandler;
export default Vue.extend({
    name: "BackToTop",
    data: () => ({
        /** Scroll distance */
        scrollTop: 0,
    }),
    computed: {
        /** Whether to display button */
        display() {
            return (this.$page.frontmatter.backToTop !== false &&
                this.scrollTop > BACK_TO_TOP_THRESHOLD);
        },
        hint() {
            return componentLocales[this.$localePath || "/"].backToTop;
        },
    },
    mounted() {
        this.scrollTop = this.getScrollTop();
        scrollHandler = debounce(() => {
            this.scrollTop = this.getScrollTop();
        }, 100);
        window.addEventListener("scroll", scrollHandler);
    },
    // eslint-disable-next-line vue/no-deprecated-destroyed-lifecycle
    beforeDestroy() {
        window.removeEventListener("scroll", scrollHandler);
    },
    methods: {
        // Get scroll distance
        getScrollTop() {
            return (window.pageYOffset ||
                document.documentElement.scrollTop ||
                document.body.scrollTop ||
                0);
        },
        // Scroll to top
        scrollToTop() {
            window.scrollTo({ top: 0, behavior: "smooth" });
            this.scrollTop = 0;
        },
    },
});
